src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java

```diff
@@ -171,47 +202,36 @@ public FilePath getScriptFile(FilePath ws) throws IOException, InterruptedExcept
             return controlDir(ws).child("script.sh");
         }
 
-        FilePath pidFile(FilePath ws) throws IOException, InterruptedException {
-            return controlDir(ws).child("pid");
-        }
-
-        private synchronized int pid(FilePath ws) throws IOException, InterruptedException {
-            if (pid == 0) {
-                FilePath pidFile = pidFile(ws);
-                if (pidFile.exists()) {
-                    try {
-                        pid = Integer.parseInt(pidFile.readToString().trim());
-                    } catch (NumberFormatException x) {
-                        throw new IOException("corrupted content in " + pidFile + ": " + x, x);
-                    }
-                }
-            }
-            return pid;
-        }
-
         @Override public Integer exitStatus(FilePath workspace, Launcher launcher) throws IOException, InterruptedException {
             Integer status = super.exitStatus(workspace, launcher);
             if (status != null) {
+                LOGGER.log(Level.FINE, "found exit code {0} in {1}", new Object[] {status, controlDir});
                 return status;
             }
-            int _pid = pid(workspace);
-            if (_pid > 0 && !ProcessLiveness.isAlive(workspace.getChannel(), _pid, launcher)) {
-                // it looks like the process has disappeared. one last check to make sure it's not a result of a race condition,
-                // then if we still don't have the exit code, use fake exit code to distinguish from 0 (success) and 1+ (observed failure)
-                // TODO would be better to have exitStatus accept a TaskListener so we could print an informative message
-                status = super.exitStatus(workspace, launcher);
-                if (status == null) {
-                    status = -1;
+            long now = System.nanoTime();
+            if (lastCheck == 0) {
+                LOGGER.log(Level.FINE, "starting check in {0}", controlDir);
+                lastCheck = now;
+            } else if (now > lastCheck + TimeUnit.SECONDS.toNanos(HEARTBEAT_CHECK_INTERVAL)) {
+                lastCheck = now;
+                long currentTimestamp = getLogFile(workspace).lastModified();
+                if (currentTimestamp == 0) {
+                    LOGGER.log(Level.FINE, "apparently never started in {0}", controlDir);
+                    return recordExitStatus(workspace, -2);
+                } else if (checkedTimestamp > 0 && currentTimestamp < checkedTimestamp + TimeUnit.SECONDS.toMillis(HEARTBEAT_MINIMUM_DELTA)) {
```

Comment:
🐛 Need to check absolute value of timestamp differences -- that way we're covered if an agent has its clock skewed and then NTP runs to reset the clocks (possible to an earlier time).  It happens more often than one would think (have seen it introduce production issues in distributed systems using timestamps). 

Now it's possible skew will cause a build to report to report under HEARTBEAT_MINIMUM even when valid but from what I've seen usually the skews are either on the order of a second or two *OR* very large -- this will cover the clock skew case much better.